### PR TITLE
Collision caching, ECS tagging, component cloning and deltatime fix

### PR DESCRIPTION
**Pull request by JanSchollaert**
# Description

- Collision caching, improved performance with the collision detector, now only calculates the collision once per entity per update.

- ECS Tags, ability to give entities tags (for further use like giving platforms a tag so a trigger won't fall through it)

- Component cloning, when creating a bullet it will clone the components. We had to implement some Rule of Five things and only implemented those we actually needed and deleted the others. This cloning was only really possible because of the ResourceManager :sweat_smile:.

[Review with](https://github.com/Brick-Studios/BeastArena/pull/32)

# Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] Any dependent changes have been merged and published in downstream modules
- [x] Included 1 or more screenshots
- [x] Code is const correct

### CODE REVIEW

File: lib/systems/physics_system.cpp [hunk 1]
```diff
@@ -18,28 +18,28 @@ void PhysicsSystem::update(double deltatime) {
         }
 
         auto transform = entityManager->getComponent<TransformComponent>(entityId);
+        // TODO make it so trigger goes through everything instead of only checking if player can move through trigger
         double mass = physics->mass;
 
         if (physics->gravity) {
-            // Also do collisions
-            double vy = physics->vy + (GRAVITY * mass);
+            double slow_down_amount = ((GRAVITY * mass) * deltatime);
```
Comment (Bram-Boris) [inline lib/systems/physics_system.cpp]: You an remove the ( and )

File: lib/systems/physics_system.cpp [hunk 2]
```diff
@@ -18,28 +18,28 @@ void PhysicsSystem::update(double deltatime) {
         }
 
         auto transform = entityManager->getComponent<TransformComponent>(entityId);
+        // TODO make it so trigger goes through everything instead of only checking if player can move through trigger
```
Comment (Bram-Boris) [inline lib/systems/physics_system.cpp]: This is fixed when we do the AABB thing?

File: include/brickengine/entities/entity_manager.hpp [hunk 3]
```diff
@@ -41,47 +43,88 @@ class EntityManager{
...
+        components_by_class->at(component_type).insert_or_assign(entity_id, std::move(component));
     }
 
     void removeEntity(const int entity_id) {
+        auto children = getChildren(entity_id);
+        // If this entity has children
+        if (!children.empty()) {
+            // Forget that you have children
+            for (auto& child_id : children) {
+                family_hierarcy_parents.erase(child_id);
```
[the remark this answers is not in the document]
Comment (JanSchollaert) [reply include/brickengine/entities/entity_manager.hpp]: No the child will not be deleted, this just deletes it from the parent hierarchy map (we misspelled this btw lmao)

File: include/brickengine/input.hpp [hunk 4]
```diff
@@ -40,20 +40,20 @@ class BrickInput {
...
+        if(inputs[player_id].count(input))
+            return inputs[player_id].at(input);
         return false;
     }
 
-    bool remapInput(int playerId, T const input) {
+    bool remapInput(int player_id, T const input) {
         //Search for the old value
-        auto oldInput = std::find_if(input_mapping.at(playerId).begin(),input_mapping.at(playerId).end(),
+        auto oldInput = std::find_if(input_mapping.at(player_id).begin(),input_mapping.at(player_id).end(),
```
[the remark this answers is not in the document]
Comment (JanSchollaert) [reply include/brickengine/input.hpp]: No we changed playerId in the playercomponent with rename symbol so thats why this is changed

File: lib/systems/physics_system.cpp (hunk 2 above)
Comment (Peter-JanGootzen) [reply lib/systems/physics_system.cpp]: Yes.

File: include/brickengine/input.hpp (hunk 4 above)
[the remark this answers is not in the document]
Comment (markJavab) [reply include/brickengine/input.hpp]: I understand, thanks for still updating it though. Continuing on the road to full snake-casing 👍

File: include/brickengine/entities/entity_manager.hpp (hunk 3 above)
[the remark this answers is not in the document]
Comment (markJavab) [reply include/brickengine/entities/entity_manager.hpp]: But doesn't that mean that the child will not have physics and stay in the air, as no gravity is applied and it's x and y are not modified?

File: include/brickengine/input.hpp (hunk 4 above)
[the remark this answers is not in the document]
Comment (Peter-JanGootzen) [reply include/brickengine/input.hpp]: Fixed it.

File: include/brickengine/entities/entity_manager.hpp (hunk 3 above)
[the remark this answers is not in the document]
Comment (Peter-JanGootzen) [reply include/brickengine/entities/entity_manager.hpp]: Fixed the misspelling.

[the remark this answers is not in the document]
Comment (markJavab) [reply include/brickengine/entities/entity_manager.hpp]: Nevermind about my last comment, I forgot about the actions that the "move out of parents house" method does.

**Pull request merged**